src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,631 @@
+     * the specified provider and is initialized with the specified parameters.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no additional parameters are provided
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code KDFParameters} is an invalid value
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param kdfParameterSpec
```

Comment:
I think this parameter should be renamed to `derivationParameterSpec` (or something similar) to avoid confusion with the `kdfParameterSpec` variable passed to `getInstance`.

Reply:
resolved.